sensus/sensus/code.py

```diff
@@ -1,28 +1,91 @@
+    # This specific function will compute the output of the neural network. 
+    
     def output(self, a):
         return sigmoid(activation(a))
     
+    ## The following functions will enable the output function to give a correct value. Our goal is to find the weights
+    ## and the biases that will lead to the correct output. To find this, we need to find how much the cost function 
+    ## (i.e., the difference between the network's output and the correct output, as provided by the training data) 
+    ## would change based on changes in the weights and the biases. 
+    
+    # We start by finding the change in the cost function based on the changes in the output activations. 
+    # How less wrong would we be if the final output activations were different? This information is the first 
+    # step towards finding the change in the cost function for weights and biases, because it will be communicated
+    # back to the previous layers by 'backpropagation'. 
+    
+    def cost_output_derivative(self, output_activations, y): # output_activations are created in backpropagation
+        """ Returns the partial derivates of the cost function with respect to all the output activations. """
+        return output_activations - y
+    
+    def sigmoid_prime(z): # multiplying this with the derivative of the cost function we get an error term, which 
+                          # tells us how much the output would change depending on the output activations. 
+        """ Returns the derivative of the cost function. """
+        return sigmoid(z) * (1-sigmoid(z))
+    
+    def backpropagation(self, x, y):
+        """ Returns the partial derivative of the cost function with respect to the weights and the biases. 
+        The output is a tuple (grad_w, grad_b). The first element is a list of vectors, each vector containing 
+        the partial derivatives of the cost function with respect to the weights in a single layer. The second 
+        is the same with respect to biases. """
+    
+    grad_w = [np.zeros(w.shape) for w in self.weights]
+    grad_b = [np.zeros(b.shape) for b in self.biases]
+    
+    # first the information (the activations / outputs) is propagated forward via the weights and the biases 
+    activation = x
+    activations = [x] # a list of vectors that will store all the activations in all the layers 
+    zs = [] # the same for the weighted sums (the activation of an input neuron which has not been into the sigmoid yet)
+    for w, b in zip(self.biases, self.weights):
+        z = np.dot(w, activation) + b
+        zs.append(z)
+        activation = sigmoid(z)
+        activations.append(activation)
+    
+    # propagating backwards: computing the error term
+    # calculated from the last layer, so index = -1
+    error = self.cost_output_derivative(activations[-1], y) * sigmoid_primte(zs[-1])
+    grad_w[-1] = np.dot(error, activations[-2].T)
```

Comment:
I think primte is a spelling mistake. Additionally, add self. for sigmoid:
self.sigmoid_prime(zs[-1])